JOSS paper

steven-murray opened the pull request:
This is the JOSS paper for 21cmFAST v3. 

Actual paper text is in `paper.md`, references are in `paper.bib` (not added yet).

Will wait for v3 release before submission.

steven-murray commented:
I have added a lot of people to be reviewers on this, because I think it concerns everyone. I'm happy for anyone who has worked on 21cmFAST to put their name and affiliation in a comment here, and I'll add it to the paper.

I'm especially looking for a read-over from @BradGreig and @andreimesinger for high-level critique. 

An example JOSS article can be found here: https://joss.theoj.org/papers/10.21105/joss.01397

andreimesinger commented:
cool.  i will go through it tuesday afternoon, after my class.

> On Mar 1, 2020, at 01:04, Steven Murray <notifications@github.com> wrote:
> 
> I have added a lot of people to be reviewers on this, because I think it concerns everyone. I'm happy for anyone who has worked on 21cmFAST to put their name and affiliation in a comment here, and I'll add it to the paper.
> 
> I'm especially looking for a read-over from @BradGreig <https://github.com/BradGreig> and @andreimesinger <https://github.com/andreimesinger> for high-level critique.
> 
> An example JOSS article can be found here: https://joss.theoj.org/papers/10.21105/joss.01397 <https://joss.theoj.org/papers/10.21105/joss.01397>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/21cmfast/21cmFAST/pull/105?email_source=notifications&email_token=ADH2UATGJG4KFAXJRD5IZ2LRFGQ7PA5CNFSM4K67MTRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENMJKXI#issuecomment-593007965>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADH2UAWY7PX2QVZRVHWILRDRFGQ7PANCNFSM4K67MTRA>.
>

BradGreig commented:
Overall this reads well and makes sense to me. What is the level of assumed physics that goes into a JOSS paper? There are a few concepts that are discussed without explanation, for which I think should be expanded on. For example, 21cm, co-eval, lightcone...

I don't think the original 21cmFAST was difficult to use. Modifying might have been more problematic but for the vast majority of people it was fairly straightforward to produce outputs they needed.

Happy to provide more comments and feedback once I get a handle on levels of assumed knowledge of key concepts etc.

JulianBMunoz commented:
Thanks for putting this together Steven!

Like Brad, I am unfamiliar with the JOSS style. All the papers there look
really short, so I guess there isn't much space to get into details. Is
this something you are thinking of submitting to the arXiv as well? If so
we might want to expand on the physics and show some examples maybe.

The only minor comment I have is either use all British or all American
spelling, we have "revolutionise" but also "reionization". Otherwise it
looks good to me!

The @munoz2018 reference for the velocities should be:
https://arxiv.org/abs/1904.07868
(and if you want to add another one:
https://arxiv.org/abs/1904.07881)

My ORCID is:
0000-0002-8984-0465
and my address, etc:
Julian B. Mu\~noz
Department of Physics, Harvard University, 17 Oxford St., Cambridge, MA
02138, USA



On Sun, Mar 1, 2020 at 8:36 PM BradGreig <notifications@github.com> wrote:

> Overall this reads well and makes sense to me. What is the level of
> assumed physics that goes into a JOSS paper? There are a few concepts that
> are discussed without explanation, for which I think should be expanded on.
> For example, 21cm, co-eval, lightcone...
>
> I don't think the original 21cmFAST was difficult to use. Modifying might
> have been more problematic but for the vast majority of people it was
> fairly straightforward to produce outputs they needed.
>
> Happy to provide more comments and feedback once I get a handle on levels
> of assumed knowledge of key concepts etc.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/21cmfast/21cmFAST/pull/105?email_source=notifications&email_token=AFLFE2IL4WAC3WUXJ2QDU23RFMESXA5CNFSM4K67MTRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENNTUBA#issuecomment-593181188>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFLFE2OXAZMMMRQVI7YEWBLRFMESXANCNFSM4K67MTRA>
> .
>

steven-murray commented:
@BradGreig  -- yes I should probably define those terms. I think the assumed level of physics knowledge for JOSS is roughly zero (it's a software journal after all). I also think you're right -- I really meant that it was hard to modify, not to use (at least, not to use as intended). I'll fix that.

@JulianBMunoz thanks for your details, I'll add them in. Also will fix the American/British conventions (I guess JOSS is US-based). I've been hopeless at this since I moved to the US. 

The point of JOSS is to have a very short paper: the philosophy is that the code itself should be the paper (by virtue of being well-documented), and that the JOSS part should merely motivate the need for the code, and put it in context with other similar codes, if applicable. 

I'd definitely submit to arXiv. Perhaps we should try putting in one or two images of lightcones or global history etc., to give a quick visual indication of what 21cmFAST is all about? 

If we want to write a paper that expands on the physics more and really defines how 21cmFAST works, we should probably be writing for Astronomy and Computing (which will take a fair bit more effort).

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+---
+title: '21cmFAST v3: A fully Python-integrated simulation code to generate 3D 21cm fluctuation fields through cosmic history.'
```
maybe something like: 
21cmFAST v3: A Python-integrated C code for generating 3D realizations of the cosmic 21cm signal and its component fields
or maybe without component fields

steven-murray replied:
Yup, I like that better (without the component fields).

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+    orcid: xxxxxx
+    affiliation: 2
+  - name: Andrei Mesinger
+    orcid: xxxxxx
```
my orcid: 0000-0003-3374-1772
affiliation: Scuola Normale Superiore, Piazza dei Cavalieri 7, 56126 Pisa, Italy

steven-murray replied:
Thanks.

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+is mapped over large swathes of the Universe's history -- has developed radically over
+the last decade.
+Notwithstanding the overwhelming technical challenges associated with the observation
+and eventual detection of this signal, the promise of the field is to revolutionise our
```
cut Notwithstanding... clause

steven-murray replied:
OK got rid of it.

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+the last decade.
+Notwithstanding the overwhelming technical challenges associated with the observation
+and eventual detection of this signal, the promise of the field is to revolutionise our
+knowledge of the first light sources: a worthy prize indeed.
```
a worthy prize indeed -> through the timing and patterns they imprint in the cosmic 21-cm signal

first light sources -> first stars, galaxies, and black holes

steven-murray replied:
Cool, did that.

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+through to fully hydrodynamical simulations of the 3D evolution of the brightness
+temperature of the spectral line.
+Between these extremes lies an especially versatile middle-ground: fast semi-analytic
+models that approximate the full 3D evolution of the field.
```
field - > relevant fields: density, velocity, temperature, ionization, and radiation.

steven-murray replied:
definitely like that better.

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+Between these extremes lies an especially versatile middle-ground: fast semi-analytic
+models that approximate the full 3D evolution of the field.
+These have the advantage of being comparable to the full first-principles
+hydrodynamic simulations, but significantly quicker to run; so much so that they can
```
comparable on the scales observable by upcoming telescopes

steven-murray replied:
good point.

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+
+
+Amongst practioners in the field of 21cm cosmology, the `21cmFAST` program has become
+the *de facto* standard for such semi-analytic simulators. `21cmFAST` [@mesinger2010]
```
please cite Mesinger and Furlanetto 2007 as well

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+the *de facto* standard for such semi-analytic simulators. `21cmFAST` [@mesinger2010]
+is a high-performance C code that uses the excursion set formalism [@furlanetto2004] to
+identify regions of ionized hydrogen atop a cosmological density field evolved using
+first- or second-order perturbation theory.
```
Lagrangian perturbation theory (you can cite e.g. zel'dovich 1970 here and Scocciomarro and Sheth 2002

The code also tracks the thermal and ionization state of the intergalactic medium, and computes X-ray, soft UV and ionizing UV cosmic radiation fields based on parametrized galaxy models.

steven-murray replied:
Thanks, added it in.

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+identify regions of ionized hydrogen atop a cosmological density field evolved using
+first- or second-order perturbation theory.
+However, `21cmFAST` is a highly specialized code, and as such has suffered from its
+difficulty of use.
```
i don't think it is difficult to use...  i would re-phrase this and below a bit to say it is not modular in structure and development..  no python support, etc.

steven-murray replied:
I make this focus more on the non-modularity now: 
```
However, `21cmFAST` is a highly specialized code, and its implementation has been
quite specific and relatively inflexible.
This inflexibility makes it hard to modify the behaviour of the code without detailed
knowledge of the full system, and without disrupting the workings of the full system.
```

andreimesinger replied:
ok

qyx268 replied:
hah, I was gonna point this out too. It's much better now

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+well-tested, stable simulator for the community.
+
+
+In `21cmFAST` v3+, which this paper presents, this goal has come a significant way towards
```
phrasing...  e.g. This paper presents v3, which enables many of the above mentioned features...  or something

steven-murray replied:
Yup, changed to `This paper presents `21cmFAST` v3+, which enables these essential new features.`

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+  previously processed data (using HDF5).
+* Simple high-level API to generate either coeval cubes or full lightcone data.
+* Improved exception handling and debugging.
+* OpenMP parallel processing.
```
OpenMP parallelization was in 21cmfast for years...

steven-murray replied:
Yeah, I guess it's something we took out then put back in.

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+* Strict `semantic versioning <https://semver.org>`_.
+
+While in v3 we have focused on the establishment of a stable and extendible infrastructure,
+we have also incorporated several new scientific features:
```
appearing in science articles

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+
+* Generate transfer functions using the `CLASS` Boltzmann code.
+* Simulate the effects of relative velocities between dark matter and Baryons [@munoz2018?].
+* Incorporation of the full mass-dependent parameterization of [@park2018].
```
park was in the previous release

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+* Generate transfer functions using the `CLASS` Boltzmann code.
+* Simulate the effects of relative velocities between dark matter and Baryons [@munoz2018?].
+* Incorporation of the full mass-dependent parameterization of [@park2018].
+* Calculate luminosity functions from the input astrophysical parameters [@xxx?]
```
again, previous release

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+* Simulate the effects of relative velocities between dark matter and Baryons [@munoz2018?].
+* Incorporation of the full mass-dependent parameterization of [@park2018].
+* Calculate luminosity functions from the input astrophysical parameters [@xxx?]
+* Mitigation of non-conservation of photons inherent in the excursion set formalism.
```
correction for non-conservation of ionizing photon (Park, Greig et al in prep)

andreimesinger commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,123 @@
+* Incorporation of the full mass-dependent parameterization of [@park2018].
+* Calculate luminosity functions from the input astrophysical parameters [@xxx?]
+* Mitigation of non-conservation of photons inherent in the excursion set formalism.
+* Mini-halo populations.
```
including molecularly-cooled galaxies with distinct properties (Qin et al. 2020)

qyx268 commented:
This is very nice! I only have some minor points. 

1. I think in most cases when we talk about 21cmFAST, we say it's a semi-numerical code. Semi-analytic model makes people think it's a galaxy-formation-type model...
2. is this just for 21cmFAST or for 21cmFAST collaboration? if it's later, I think we could say something about 21CMMC as well? 
3. for the new feature, if 21CMMC is considered, we can talk a bit more about forward-modelling of the EE PS using CLASS and inclusion of the Planck 2018 likelihood using clik. I think they are almost ready to be merged into master. I'm writing a draft of it and it shouldn't be too hard to make a PR. We just need to clear a bit of its writing...

steven-murray commented:
Thanks @qyx268. OK I will change nomenclature to "semi-numerical".

This is specifically for `21cmFAST`, not 21CMMC -- we will probably do another short paper for that when it's ready (hopefully in the next month or two).

steven-murray commented:
I've addressed all comments that people have made so far. @qyx268 I think it would be good for you to be on the paper: can you give me your details (orcid)?

I've added a new section: Example, where I show the simplest possible example of how to make a lightcone, trying to show off some of the cool stuff it can do with the least code lines. Let me know if you think this is a good example!

BradGreig commented:
@steven-murray, I realised I didn't provide affiliation/orcid ID stuff (was focusing on other aspects).

ORCID: 0000-0002-4085-2094
Affiliations:
1. ARC Centre of Excellence for All-Sky Astrophysics in 3 Dimensions (ASTRO 3D)
2. School of Physics, University of Melbourne, Parkville, VIC 3010, Australia

JulianBMunoz commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,202 @@
+However, `21cmFAST` is a highly specialized code, and its implementation has been
+quite specific and relatively inflexible.
+This inflexibility makes it hard to modify the behaviour of the code without detailed
+knowledge of the full system, and without disrupting the workings of the full system.
```
full system is repeated, change the last one to code maybe?

steven-murray replied:
Thanks, changed to "This inflexibility makes it difficult to modify the behaviour of the code without detailed knowledge of the full system, or disrupting its workings."

JulianBMunoz commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,202 @@
+
+# Summary
+
+The field of 21cm cosmology -- in which the hyperfine spectral line of neutral hydrogen
```
I would hyphenate this 21-cm, and put a space in the next one. I don't think 21 and cm should go together (other than in the name of the code :)

steven-murray replied:
Awesome, yup makes sense.

JulianBMunoz commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,202 @@
+* Generate transfer functions using the `CLASS` Boltzmann code.
+* Simulate the effects of relative velocities between dark matter and Baryons [@munoz2019a,@munoz2019b].
+* Correction for non-conservation of ionizing photons (Park, Greig et al., *in prep*).
+* Including molecularly-cooled galaxies with distinct properties [@qin2020]
```
No hyphen here, since molercularly ends in ly.

JulianBMunoz commented on joss-paper/Makefile:
```diff
@@ -0,0 +1,58 @@
+issue   := 28
+gh_issue  := 850
+year    := 2018
+submitted := 23 July $(year)
```
Should we change this? Or it doesn't matter?

steven-murray replied:
I'll change this when it's closer to submission, so I actually know the date :-) Note: the Makefile is not actually part of the submission, it's just a way for us to make a PDF from the markdown (which happens automatically in the JOSS review) to eventually submit to arXiv.

jaehongpark00 commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,202 @@
+# Summary
+
+The field of 21cm cosmology -- in which the hyperfine spectral line of neutral hydrogen
+(appearing at rest-frame wavelengths of 21cm) is mapped over large swathes of the
```
I think it should be "the rest-frame wavelength of 21cm". ???

steven-murray replied:
Yes, you're right, I've now fixed it!

jaehongpark00 commented on joss-paper/paper.md:
```diff
@@ -0,0 +1,202 @@
+
+While in v3 we have focused on the establishment of a stable and extendable infrastructure,
+we have also incorporated several new scientific features, appearing in separate papers:
+
```
It would be good to add 
+ * Calculate the rest-frame UV luminosity functions based on the parametrized galaxy models.

I'm not sure it is good to put this here. But, the previous 21cmFAST didn't include this calculation, it was only in 21CMMC. Anyway, I think this feature should be pointed out somewhere.

steven-murray replied:
Thanks @jaehongpark00, I'll add that in.

qyx268 commented:
Thanks @steven-murray  
ORCID: 0000-0002-4314-1810
Affiliation: Scuola Normale Superiore, Piazza dei Cavalieri 7, I-56126 Pisa, Italy

jaehongpark00 commented:
@steven-murray Thank you for this nice draft.
This looks good to me.

My ORCID id is
0000-0003-3095-6137
Affiliation: 4

qyx268 commented:
btw, the links to the plots don't seem to work at this moment.

steven-murray commented:
@qyx yes the plots have required the new PR #108, which means they aren't in this PR. Once that has been merged into master, I'll pull it into this branch and the plots will be available.

qyx268 reviewed (approved):
looks all good!
